network/native_endpointclient_linux.go

```diff
@@ -0,0 +1,403 @@
+		return errors.Wrap(err, "failed adding routes to vnet specific to this container")
+	}
+	// Return to ConfigureContainerInterfacesAndRoutes
+	return err
+}
+
+// Helper that gets the routes in the vnet NS for a particular list of IP addresses
+// Example: 192.168.0.4 dev <device which connects to NS with that IP> proto static
+func (client *NativeEndpointClient) GetVnetRoutes(ipAddresses []net.IPNet) []RouteInfo {
+	routeInfoList := make([]RouteInfo, 0, len(ipAddresses))
+	// Add route specifying which device the pod ip(s) are on
+	for _, ipAddr := range ipAddresses {
+		var (
+			routeInfo RouteInfo
+			ipNet     net.IPNet
+		)
+
+		if ipAddr.IP.To4() != nil {
+			ipNet = net.IPNet{IP: ipAddr.IP, Mask: net.CIDRMask(ipv4FullMask, ipv4Bits)}
+		} else {
+			ipNet = net.IPNet{IP: ipAddr.IP, Mask: net.CIDRMask(ipv6FullMask, ipv6Bits)}
+		}
+		log.Printf("[net] Native client adding route for the ip %v", ipNet.String())
+		routeInfo.Dst = ipNet
+		routeInfoList = append(routeInfoList, routeInfo)
+
+	}
+	return routeInfoList
+}
+
+// Helper that creates routing rules for the current NS which direct packets
+// to the virtual gateway ip on linkToName device interface
+// Route 1: 169.254.1.1 dev <linkToName>
+// Route 2: default via 169.254.1.1 dev <linkToName>
+func (client *NativeEndpointClient) AddDefaultRoutes(linkToName string) error {
+	// Add route for virtualgwip (ip route add 169.254.1.1/32 dev eth0)
+	virtualGwIP, virtualGwNet, _ := net.ParseCIDR(virtualGwIPString)
+	routeInfo := RouteInfo{
+		Dst:   *virtualGwNet,
+		Scope: netlink.RT_SCOPE_LINK,
+	}
+	// Difference between interface name in addRoutes and DevName: in RouteInfo?
+	if err := addRoutes(client.netlink, client.netioshim, linkToName, []RouteInfo{routeInfo}); err != nil {
+		return err
+	}
+
+	// Add default route (ip route add default via 169.254.1.1 dev eth0)
+	_, defaultIPNet, _ := net.ParseCIDR(defaultGwCidr)
+	dstIP := net.IPNet{IP: net.ParseIP(defaultGw), Mask: defaultIPNet.Mask}
+	routeInfo = RouteInfo{
+		Dst: dstIP,
+		Gw:  virtualGwIP,
+	}
+
+	if err := addRoutes(client.netlink, client.netioshim, linkToName, []RouteInfo{routeInfo}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Helper that creates arp entry for the current NS which maps the virtual
+// gateway (169.254.1.1) to destMac on a particular interfaceName
+// Example: (169.254.1.1) at 12:34:56:78:9a:bc [ether] PERM on <interfaceName>
+func (client *NativeEndpointClient) AddDefaultArp(interfaceName, destMac string) error {
+	_, virtualGwNet, _ := net.ParseCIDR(virtualGwIPString)
+	log.Printf("[net] Adding static arp for IP address %v and MAC %v in namespace",
+		virtualGwNet.String(), destMac)
+	hardwareAddr, err := net.ParseMAC(destMac)
+	if err != nil {
+		return errors.Wrap(err, "unable to parse mac")
+	}
+	if err := client.netlink.AddOrRemoveStaticArp(netlink.ADD,
+		interfaceName,
+		virtualGwNet.IP,
+		hardwareAddr,
+		false); err != nil {
```

Comment:
unwrap
splitting method signatures or invocations across lines is worse for readability